Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -0,0 +1,159 @@
+
+    /// The state of the HTTP connection.
+    private enum ConnectionState {
+        /// We are waiting for a HTTP response to complete before we
+        /// let the next request in.
+        case responseEndPending
+
+        /// Neither a request nor a response are complete. This can be either
+        /// because nothing is happening on the connection, or because the request
+        /// and response are simultaneously progressing and have not completed.
+        case idle
+
+        /// The server has responded early, before the request has completed. We need
+        /// to wait for the request to complete, but won't block anything.
+        case requestEndPending
+
+        mutating func responseEndReceived() {
+            switch self {
+            case .responseEndPending:
+                // Got the response we were waiting for.
+                self = .idle
+            case .idle:
+                // We got a response while still receiving a request, which we have to
+                // wait for.
+                self = .requestEndPending
+            case .requestEndPending:
+                preconditionFailure("Received second response")
+            }
+        }
+
+        mutating func requestEndReceived() {
+            switch self {
+            case .requestEndPending:
+                // Got the request end we were waiting for.
+                self = .idle
+            case .idle:
+                // We got a request and the response isn't done, wait for the
+                // response.
+                self = .responseEndPending
+            case .responseEndPending:
+                preconditionFailure("Received second request")
+            }
+        }
+    }
+
+    /// The connection state
+    private var state = ConnectionState.idle
+
+    /// While we're waiting to send the response we don't read from the socket.
+    /// This keeps track of whether we need to call read() when we've send our response.
+    private var readPending = false
+
+    /// The buffered HTTP requests that are not going to be addressed yet. In general clients
+    /// don't pipeline, so this initially allocates no space for data at all. Clients that
+    /// do pipeline will cause dynamic resizing of the buffer, which is generally acceptable.
+    private var requestBuffer = CircularBuffer<NIOAny>(initialRingCapacity: 0)
+
+    public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        if case .responseEndPending = self.state {
+            self.requestBuffer.append(data)
+            return
+        }
+
+        if case .end = self.unwrapInboundIn(data) {
+            // New request is complete. We don't want any more data from now on.
+            self.state.requestEndReceived()
+        }
+        ctx.fireChannelRead(data)
+    }
+
+    public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        assert(self.state != .requestEndPending, "Unexpected data flow")
```

Comment:
if I see this correctly, this happens when a second response is coming in end we're waiting for the previous request's `.end`. Could we have a better message here?